File: sdk/go/hulo.go
package hulo

import context "context"

var _ HuloInterfaceServer = &HuloInterface{}

type (
	HuloCall       func(req *CallRequest) (*CallResponse, error)
	HuloCompletion func(req *CompletionRequest) (*CompletionResponse, error)
)

type HuloInterface struct {
	UnimplementedHuloInterfaceServer
	dict      map[string]HuloCall
	completer HuloCompletion
}

func NewHuloInterface() *HuloInterface {
	return &HuloInterface{
		dict: make(map[string]HuloCall),
		completer: func(req *CompletionRequest) (*CompletionResponse, error) {
			return &CompletionResponse{}, nil
		},
	}
}

func (hi *HuloInterface) Call(ctx context.Context, req *CallRequest) (*CallResponse, error) {
	if call, ok := hi.dict[req.Func]; ok {
		return call(req)
	}
	return &CallResponse{Buf: "unknown"}, nil
}

func (hi *HuloInterface) Completion(ctx context.Context, req *CompletionRequest) (*CompletionResponse, error) {
	return hi.completer(req)
}

func (hi *HuloInterface) Register(name string, call HuloCall) {
	hi.dict[name] = call
}

func (hi *HuloInterface) UnRegister(name string) {
	delete(hi.dict, name)
}

File: sdk/dart/hulo.dart
import 'dart:async';
import 'dart:io';
import 'dart:mirrors';
import 'package:args/args.dart';
import 'package:grpc/grpc.dart';
import 'hulo.pbgrpc.dart';

typedef HuloCall = Future<CallResponse> Function(ServiceCall, CallRequest);

class HuloInterface extends HuloInterfaceServiceBase {
  late Map<String, HuloCall> dict;

  HuloInterface() {
    this.dict = new Map();
  }

  @override
  Future<CallResponse> call(ServiceCall call, CallRequest request) async {
    if (this.dict.containsKey(request.func)) {
      print(this.dict);
      return this.dict[request.func]!(call, request);
    }
    return CallResponse(buf: "unknown");
  }

  @override
  Future<CompletionResponse> completion(
      ServiceCall call, CompletionRequest request) {
    throw UnimplementedError();
  }

  void Register(String name, HuloCall cb) {
    this.dict[name] = cb;
  }
}

class Call {
  final String name;
  const Call(this.name);
}

class HuloBuilder {
  late final Server server;
  late int port;

  HuloBuilder(List<String> args) {
    HuloInterface hulo = new HuloInterface();
    final parser = ArgParser();
    parser.addOption('port', abbr: 'p');
    ArgResults result = parser.parse(args);
    try {
      port = int.parse(result['port']);
    } catch (e) {
      print("invalid port");
      exit(1);
    }
    currentMirrorSystem().libraries.forEach((_, lib) {
      lib.declarations.forEach((s, decl) {
        decl.metadata.where((m) => m.reflectee is Call).forEach((m) {
          var anno = m.reflectee as Call;
          if (decl is MethodMirror) {
            hulo.Register(anno.name, (ServiceCall call, CallRequest req) async {
              return ((decl).owner as LibraryMirror)
                  .invoke(s, [call, req]).reflectee;
            });
          }
          ;
        });
      });
    });
    server = Server([hulo]);
  }

  Future<void> Start() async {
    print("service start...");
    await this.server.serve(port: this.port);
  }
}

class HuloClient {
  late final ClientChannel channel;
  late final HuloInterfaceClient cli;

  HuloClient(int port) {
    this.channel = ClientChannel('localhost',
        port: port,
        options:
            const ChannelOptions(credentials: ChannelCredentials.insecure()));
    this.cli = HuloInterfaceClient(channel);
  }

  Future<CallResponse> Call(CallRequest request) {
    return this.cli.call(request);
  }

  Future<void> close() async {
    this.channel.shutdown();
  }
}
